first version of causync

causalitygroup opened the pull request.

causalitygroup committed:
first version of causync

gabalz commented on causync.py:
```diff
@@ -0,0 +1,291 @@
+
+class CauSync(object):
+    """ CauSync object for Sync-related functions. """
+    def __init__(self,
```
Documentation of these parameters describing their role would be nice to have.

gabalz commented on causync.py:
```diff
@@ -0,0 +1,291 @@
+        return result
+
+    def run_cleanup(self, incremental=False):
+        """ This function is executed when the task argument is 'cleanup'. """
```
What does a cleanup do? What is the role of the incremental flag?

gabalz commented on causync.py:
```diff
@@ -0,0 +1,291 @@
+    arguments.checkonly = True if arguments.task == 'check' else False
+    arguments.cleanup = True if arguments.task == 'cleanup' else False
+
+    if arguments.incremental and not arguments.incremental_basedir:
```
Incremental backup would be nice to have in an automatic way if previous backup directories exist.

gabalz commented on causync.py:
```diff
@@ -0,0 +1,291 @@
+                                                              d=curdate))
+
+        if self.incremental and self.incremental_basedir:
+            extra_flags = "--link-dest={}".format(self.incremental_basedir)
```
It would be better to use multiple directories for previous backups, so we get incrementality even if a few days are missing.

causalitygroup commented on causync.py:
```diff
@@ -0,0 +1,291 @@
+
+        self.logger.info("successfully deleted old backups")
+
+    def create_lockfile(self):
```
@gabalz, our initial specification mentions a lock directory, but this creates a lock file instead. It feels safer and easier to find (for example with `find . -type f`). Is this a good/better solution, or should I change it to a directory?

gabalz replied:
Hmm, I don't see any problem with a lock file, please keep it for now.

causalitygroup committed:
added docstrings, improved incremental backups, testable functions

incremental backups: find basedir automatically when no
--incremental-basedir is supplied
testable functions: added separate functions for working with
directories, backup dates and the lockfile

causalitygroup committed:
added nosetests for dates, directory names and lockfile

causalitygroup committed:
removed incremental args, added date-interval arg, added find function

- incremental: removed --incremental and --incremental basedir, added
--no-incremental instead (we don't have a real basedir because of hard
links anyway)
- date-interval: backups are separated to yearly, monthly, mondays and
daily, so cleanup is easier. Daily is the default.
- find function: finds latest backups to use with --link-dest, finds
oldest backups for cleanup
- added tests for finding new and old backups
- removed test for incremental directories

causalitygroup committed:
updated backup destination directory names, improved find_old_backups()

causalitygroup committed:
".idea/" added to .gitignore

causalitygroup committed:
added logging for rmtree(), removed result line from is_running()

gabalz commented on tests/test_backups.py:
```diff
@@ -0,0 +1,57 @@
+
+def test_find_old_backups():
+    cs = CauSync(config, "/tmp/causync_src", "/tmp/causync_dest")
+    cs.config.DATE_FORMAT = "%y%m%d_%H%M%S"
```
Can we change the format to date only: %Y%m%d, even in the tests?
If causync is re-run on the same day, can we update the previous backup instead of creating a new one?

susmus replied:
Yes, I'll update the format in the config and test it. Rsync should handle updating when we re-run it, but I'll also test that.

causalitygroup replied:
Yes, I'll change the date and test it. Rsync should handle updating automatically on re-runs, but I'll also test that.

gabalz commented on tests/test_backups.py:
```diff
@@ -0,0 +1,57 @@
+    cs = CauSync(config, "/tmp/causync_src", "/tmp/causync_dest")
+    cs.config.DATE_FORMAT = "%y%m%d_%H%M%S"
+
+    dirnames = [
```
Can we expand this test to make sure that the weekly, monthly, yearly backup keeping functionality works?

gabalz commented on config.py:
```diff
@@ -0,0 +1,30 @@
+
+# how many backups should we get
+BACKUPS_LINK_DEST_COUNT=5
+BACKUPS_TO_KEEP = {
```
Please use yearly:10, monthly:6, weekly:4, daily:7.

gabalz commented on causync.py:
```diff
@@ -0,0 +1,374 @@
+        dirdates.sort(reverse=True)
+        keep.sort(reverse=True)
+
+        return keep, delete
```
What is the purpose of calculating and returning keep?
You do not seem to use it anywhere...

gabalz commented on causync.py:
```diff
@@ -0,0 +1,374 @@
+        monthly = self.find_old_backups(listdir, 'monthly', self.config.BACKUPS_TO_KEEP['monthly'])
+        weekly = self.find_old_backups(listdir, 'weekly', self.config.BACKUPS_TO_KEEP['weekly'])
+        daily = self.find_old_backups(listdir, 'daily', self.config.BACKUPS_TO_KEEP['daily'])
+
```
This looks quite chaotic, you simply need the aggregated list of delete: sorted(set(yearly[1] + monthly[1] + weekly[1] + daily[1])). Then you can delete them all...

gabalz commented on causync.py:
```diff
@@ -0,0 +1,374 @@
+        weekly = [weekly[0], set(weekly[1]) - set(yearly[0]) - set(monthly[0])]
+        daily = [daily[0], set(daily[1]) - set(monthly[0]) - set(weekly[0]) - set(yearly[0])]
+
+        self.rmtree(list(sorted(yearly[1])))
```
Please log which directories you delete.

causalitygroup committed:
added comments and test to find_old_backups(), updated numbers in config

gabalz commented on causync.py:
```diff
@@ -0,0 +1,401 @@
+        weekly = [weekly[0], set(weekly[1]) - set(yearly[0]) - set(monthly[0])]
+        daily = [daily[0], set(daily[1]) - set(monthly[0]) - set(weekly[0]) - set(yearly[0])]
+
+        d = set(list(yearly[1]) + list(monthly[1]) + list(weekly[1]) + list(daily[1]))
```
Can we have instead of this code that `d = set(listdir) - set(yearly[0]) - set(monthly[0]) - set(daily[0])`? Is this correct?

causalitygroup committed:
readme added
